crates/abd-clam/src/core/graph/_graph.rs

```diff
@@ -760,9 +748,138 @@ impl<'a, U: Number> Graph<'a, U> {
+        let name = "test".to_string();
+        VecDataset::new(name, data, metric, false)
+    }
+    /// Euclidean distance between two vectors.
+    pub fn euclidean<T: Number, F: Float>(x: &Vec<T>, y: &Vec<T>) -> F {
+        distances::vectors::euclidean(x, y)
+    }
+    #[test]
+    fn create_graph() {
+        let data = gen_dataset(1000, 10, 42, euclidean);
+        let partition_criteria: PartitionCriteria<f32> = PartitionCriteria::default();
+        let raw_tree = Tree::new(data, Some(42)).partition(&partition_criteria);
+        let selected_clusters = select_clusters(
+            raw_tree.root(),
+            &pretrained_models::get_meta_ml_scorers().first().unwrap().1,
+            4,
+        );
+
+        let edges = detect_edges(&selected_clusters, raw_tree.data());
+
+        let graph = Graph::from_clusters_and_edges(selected_clusters.clone(), edges.clone());
+        assert!(graph.is_ok());
+        if let Ok(graph) = graph {
+            // assert edges and clusters are correct
+            assert_eq!(graph.clusters().len(), selected_clusters.len());
+            assert_eq!(graph.edges().len(), edges.len());
+
+            let reference_population = selected_clusters.iter().fold(0, |acc, &c| acc + c.cardinality());
+            assert_eq!(graph.population(), reference_population);
+            let components = graph.find_component_clusters();
+
+            // assert ordered clusters are in correct order
+            graph.clusters().iter().for_each(|c1| {
+                assert!(graph.ordered_clusters.contains(c1));
+            });
+
+            let num_clusters_in_components = components.iter().map(|c| c.len()).sum::<usize>();
+            assert_eq!(num_clusters_in_components, selected_clusters.len());
+
+            // assert the number of clusters in a component is equal to the number of clusters in each of its cluster's traversals
+            for component in &components {
+                for c in component {
+                    if let Ok(traversal_result) = graph.traverse(c) {
+                        assert_eq!(traversal_result.0.len(), component.len());
+                        // assert_eq!(traversal_result.1.len(), component.len());
+                    }
+                }
+            }
+        }
+    }
+
+    #[test]
+    fn adjacency_map() {
+        let data = gen_dataset(1000, 10, 42, euclidean);
+        let partition_criteria: PartitionCriteria<f32> = PartitionCriteria::default();
+        let raw_tree = Tree::new(data, Some(42)).partition(&partition_criteria);
+        let selected_clusters = select_clusters(
+            raw_tree.root(),
+            &pretrained_models::get_meta_ml_scorers().first().unwrap().1,
+            4,
+        );
+
+        let edges = detect_edges(&selected_clusters, raw_tree.data());
+
+        let graph = Graph::from_clusters_and_edges(selected_clusters.clone(), edges);
+        assert!(graph.is_ok());
```

Comment:
this is tricky because i am testing both graph::from_tree and graph::from_clusters_and_edges. I did refactor the code a bit to remove some duplicate code though.